File: kotlin_jetpack_compose/app/src/main/java/com/example/testapp/data/supabase/repository/TrainingRepositoryImpl.kt
package com.example.testapp.data.supabase.repository

import com.example.testapp.data.supabase.model.DTOs.AttendanceUpdateRequestDto
import com.example.testapp.data.supabase.model.DTOs.CreateRasporedRequestDto
import com.example.testapp.data.supabase.model.DTOs.RasporedCreateDto
import com.example.testapp.data.supabase.model.DTOs.toDomain
import com.example.testapp.data.supabase.model.mappers.availableTrainingEdgeDtoToDomain
import com.example.testapp.data.supabase.model.mappers.prijavaFullDtoToDomain
import com.example.testapp.data.supabase.model.mappers.toDomain
import com.example.testapp.data.supabase.model.mappers.toDto
import com.example.testapp.data.supabase.network.service.SupabaseTrainingService
import com.example.testapp.domain.model.AttendanceUpdate
import com.example.testapp.domain.model.AttendanceUpdateResult
import com.example.testapp.domain.model.CreateRasporedRequest
import com.example.testapp.domain.model.CreateTreningRequest
import com.example.testapp.domain.model.CreateTreningResult
import com.example.testapp.domain.model.DeleteResult
import com.example.testapp.domain.model.DostupniTrening
import com.example.testapp.domain.model.Dvorana
import com.example.testapp.domain.model.PrijavljenTrening
import com.example.testapp.domain.model.PrijavljeniSudionik
import com.example.testapp.domain.model.Raspored
import com.example.testapp.domain.model.Teretana
import com.example.testapp.domain.model.TrainingDetails
import com.example.testapp.domain.model.TrenerTrening
import com.example.testapp.domain.model.TreningOption
import com.example.testapp.domain.model.VrstaTreninga
import com.example.testapp.domain.repository.TrainingRepository
import java.time.LocalDate
import javax.inject.Inject

class TrainingRepositoryImpl @Inject constructor(
    private val supabaseTrainingService: SupabaseTrainingService
) : TrainingRepository {

    override suspend fun getTrainingsForUser(sportasId: String): List<PrijavljenTrening> {
        return supabaseTrainingService.getTrainingsForUser(sportasId)
            .map { prijavaFullDtoToDomain(it) }
    }

    override suspend fun getTrainingsByDateAndTeretana(
        teretana: String,
        date: LocalDate
    ): List<DostupniTrening> {
        return supabaseTrainingService.getAvailableTrainings(date = date, teretanaId = teretana)
            .map { availableTrainingEdgeDtoToDomain(it) }
    }

    override suspend fun prijaviSeNaTrening(rasporedId: String) {
        val korisnikId = "temp" //used just for testing
        supabaseTrainingService.signUpForTraining(
            rasporedId = rasporedId,
            korisnikId = korisnikId
        )
    }

    override suspend fun getTrainingDetails(treningId: String): TrainingDetails {
        val dto = supabaseTrainingService.getTrainingDetails(treningId)
        return dto.toDomain()
    }

    override suspend fun getDvorane(): List<Dvorana> {
        return supabaseTrainingService.getDvorane().map { it.toDomain() }
    }

    override suspend fun getTeretane(): List<Teretana> {
        return supabaseTrainingService.getTeretane().map { it.toDomain() }
    }

    override suspend fun getTrainingsForTrainer(trenerId: String): List<TrenerTrening> {
        return supabaseTrainingService.getTrainingsForTrainer(trenerId).map { it.toDomain() }
    }

    override suspend fun getAttendeesForRaspored(rasporedId: String): List<PrijavljeniSudionik> {
        return supabaseTrainingService
            .getAttendeesByRaspored(rasporedId)
            .map { it.toDomain() }
    }

    override suspend fun setAttendanceForRaspored(
        rasporedId: String,
        updates: List<AttendanceUpdate>
    ): AttendanceUpdateResult {
        val request = AttendanceUpdateRequestDto(
            rasporedId = rasporedId,
            updates = updates.map { it.toDto() }
        )

        val response = supabaseTrainingService.setAttendanceForRaspored(request)
        return response.toDomain()
    }

    override suspend fun getVrsteTreninga(): List<VrstaTreninga> {
        return supabaseTrainingService.getVrsteTreninga().map { it.toDomain() }
    }

    override suspend fun createTrening(request: CreateTreningRequest): CreateTreningResult {
        val dto = request.toDto()
        val resp = supabaseTrainingService.createTrening(dto)
        return resp.toDomain()
    }

    override suspend fun getTreningOptions(): List<TreningOption> {
        return supabaseTrainingService.getTreningOptions().map { it.toDomain() }
    }

    override suspend fun createRaspored(request: CreateRasporedRequest): Raspored {
        val reqDto = CreateRasporedRequestDto(
            raspored = RasporedCreateDto(
                IdRasporeda = request.idRasporeda,
                IdTreninga = request.treningId,
                PocetakVrijeme = request.pocetak.toString(),
                ZavrsetakVrijeme = request.zavrsetak.toString(),
                IdTrenera = request.trenerId
            )
        )

        val createdDto = supabaseTrainingService.createRaspored(reqDto)
        return createdDto.toDomain()
    }

    override suspend fun deleteRaspored(rasporedId: String): DeleteResult {
        val dto = supabaseTrainingService.deleteRaspored(rasporedId)
        return DeleteResult(success = dto.success)
    }


}

File: flutter/test_app/lib/data/supabase/repository/training_repository_impl.dart
import 'package:supabase_flutter/supabase_flutter.dart';
import 'package:test_app/data/supabase/model/DTOs/trainings_dtos.dart';

import '../../../domain/model/common_models.dart';
import '../../../domain/model/prijavljen_trening.dart';
import '../../../domain/model/training_models.dart';
import '../../../domain/model/trener_models.dart';
import '../../../domain/repository/training_repository.dart';
import '../model/DTOs/trener_dtos.dart';
import '../model/mapper/training_mapper.dart';
import '../network/service/training_service.dart';

class TrainingRepositoryImpl implements TrainingRepository {
  final TrainingService service;

  TrainingRepositoryImpl(this.service);

  @override
  Future<List<PrijavljenTrening>> getTrainingsForUser(String sportasId) async {
    try {
      final prijaveDto = await service.getPrijaveFullForUser(sportasId);
      return prijaveDto.map(prijavaFullDtoToDomain).toList();
    } catch (_) {
      return [];
    }
  }

  @override
  Future<List<Teretana>> getTeretane() async {
    final dtoList = await service.getTeretane();
    return dtoList.map((d) => d.toDomain()).toList();
  }

  @override
  Future<List<DostupniTrening>> getTrainingsByDateAndTeretana({
    required String teretanaId,
    required DateTime date,
  }) async {
    final dtoList = await service.getAvailableTrainings(
      teretanaId: teretanaId,
      date: date,
    );
    return dtoList.map((d) => d.toDomain()).toList();
  }

  @override
  Future<SignUpResult> signUpForTraining({
    required String rasporedId,
    required String korisnikId,
  }) async {
    try {
      final data = await service.signUpForTraining(
        korisnikId: korisnikId,
        rasporedId: rasporedId,
      );

      if (data != null && data['success'] == true) {
        return SignUpResult.success;
      }

      return SignUpResult.error;
    } on PostgrestException catch (e) {
      final msg = e.message ?? '';
      if (msg.contains('already')) {
        return SignUpResult.userAlreadySigned;
      }
      if (msg.contains('full')) {
        return SignUpResult.trainingFull;
      }

      return SignUpResult.error;
    } catch (_) {
      return SignUpResult.error;
    }
  }

  @override
  Future<List<TrainerTraining>> getTrainingsForTrainer(String trenerId) async {
    final List<RasporedFullDto> dtoList =
    await service.getTrainingsForTrainer(trenerId: trenerId);

    return dtoList.map(RasporedFullMapper.toTrainerTraining).toList();
  }

  @override
  Future<List<PrijavljeniSudionik>> getAttendeesForRaspored(String rasporedId) async {
    final List<AttendeeDto> dtoList =
    await service.getAttendeesByRaspored(rasporedId: rasporedId);

    return dtoList.map(TrainerMappers.toDomainAttendee).toList();
  }

  @override
  Future<AttendanceUpdateResult> setAttendanceForRaspored({
    required String rasporedId,
    required List<AttendanceUpdate> updates,
  }) async {
    final request = TrainerMappers.toAttendanceRequestDto(
      rasporedId: rasporedId,
      updates: updates,
    );

    final AttendanceUpdateResponseDto response =
    await service.setAttendanceForRaspored(request: request);

    return TrainerMappers.toDomainAttendanceResult(response);
  }

  @override
  Future<List<VrstaTreninga>> getVrsteTreninga() async {
    final dtos = await service.getVrsteTreninga();
    return dtos.map((d) => d.toDomain()).toList();
  }

  @override
  Future<List<VrstaTreningaOption>> getVrsteTreningaOptions() async {
    final dtos = await service.getVrsteTreninga();
    return dtos.map((d) => d.toOption()).toList();
  }

  @override
  Future<List<TreningOption>> getTreningOptions() async {
    final dtos = await service.getTreningOptions();
    return dtos.map((d) => d.toDomain()).toList();
  }

  @override
  Future<void> createTrening(CreateTreningInput input) async {
    final treningDto = TreningCreateDto(
      idDvOdr: input.dvoranaId,
      maksBrojSportasa: input.maksBrojSportasa,
      idVrTreninga: input.existingVrstaId,
      idLaksijegTreninga: input.idLaksijegTreninga,
      idTezegTreninga: input.idTezegTreninga,
    );

    VrstaTreningaCreateDto? vrstaDto;
    if (input.existingVrstaId == null) {
      if (input.newVrstaNaziv == null || input.newVrstaNaziv!.trim().isEmpty) {
        throw Exception('Naziv vrste treninga je obavezan.');
      }
      if (input.newVrstaTezina == null) {
        throw Exception('Težina vrste treninga je obavezna.');
      }
      vrstaDto = VrstaTreningaCreateDto(
        nazivVrTreninga: input.newVrstaNaziv!.trim(),
        tezina: input.newVrstaTezina!,
      );
    }

    final req = CreateTreningRequestDto(
      trening: treningDto,
      vrsta: vrstaDto,
    );

    await service.createTrening(req);
  }

  @override
  Future<void> createRaspored(CreateRasporedInput input) async {
    final startIso = input.start.toUtc().toIso8601String();
    final endIso = input.end.toUtc().toIso8601String();

    final req = CreateRasporedRequestDto(
      idTreninga: input.treningId,
      idTrenera: 'IGNORED',
      pocetakVrijemeIso: startIso,
      zavrsetakVrijemeIso: endIso,
    );

    final dto = await service.createRaspored(req);
    return;
  }

  @override
  Future<DeleteRasporedResult> deleteRaspored(String rasporedId) async {
    final dto = await service.deleteRaspored(rasporedId);
    return dto.toDomain();
  }

  @override
  Future<List<DvoranaSimple>> getDvorane() async {
    final dtoList = await service.getDvorane();
    return dtoList.map((d) => d.toDomain()).toList();
  }

}
